quark.c

```diff
@@ -1329,7 +1588,102 @@ sproc_namespace(struct quark_process *qp, const char *path, u32 *dst, int dfd)
 }
 
 static int
-sproc_pid(struct quark_queue *qq, int pid, int dfd)
+sproc_pid_sockets(struct quark_queue *qq,
+    struct sproc_socket_by_inode *by_inode, int pid, int dfd)
+{
+	DIR		*dir;
+	struct dirent	*d;
+	int		 fdfd;
+
+	if ((fdfd = openat(dfd, "fd", O_RDONLY)) == -1) {
+		qwarn("open fd");
+		return (-1);
+	}
+	dir = fdopendir(fdfd);
+	if (dir == NULL) {
+		close(fdfd);
+		qwarn("fdopendir fdfd");
+		return (-1);
+	}
+
+	while ((d = readdir(dir)) != NULL) {
+		char			 buf[256];
+		ssize_t			 n;
+		u_long			 inode	= 0;
+		const char		*needle = "socket:[";
+		struct sproc_socket	*ss, ss_key;
+		struct quark_socket	*qsk, qsk_key;
+
+		if (d->d_type != DT_LNK)
+			continue;
+		n = qreadlinkat(fdfd, d->d_name, buf, sizeof(buf));
+		if (n == -1)
+			qwarn("qreadlinkat %s", d->d_name);
+		if (n <= 0)
+			continue;
+		if (strncmp(buf, needle, strlen(needle)))
+			continue;
+		if (sscanf(buf, "socket:[%lu]", &inode) != 1) {
+			qwarnx("sscanf can't get inode");
+			continue;
+		}
+
+		bzero(&ss_key, sizeof(ss_key));
+		ss_key.inode = inode;
+		ss = RB_FIND(sproc_socket_by_inode, by_inode, &ss_key);
+		/*
+		 * We're only interested in TCP sockets, we end up finding
+		 * AF_UNIX and SOCK_DGRAM here as well, so it's normal to have
+		 * many misses.
+		 */
+		if (ss == NULL)
+			continue;
+
+		/*
+		 * Maybe another process already references the same socket.
+		 */
+		bzero(&qsk_key, sizeof(qsk_key));
+		qsk_key.local = ss->socket.local;
+		qsk_key.remote = ss->socket.remote;
+		qsk = RB_FIND(socket_by_src_dst, &qq->socket_by_src_dst,
+		    &qsk_key);
+		if (qsk != NULL) {
+			/* XXX TODO WRITE AN EXAMPLE TO TRIGGER THIS */
+			printf("EXISTING !!! pid %d fd %s -> %s (inode=%lu, ss=%p)\n", pid,
```

Comment:
Is this an example of a process forking without closing the socket?

Reply:
Yep, or sending the socket via sendmsg() to another process, but basically any case where a socket is referenced/used by more than one process.